### ISSUE TITLE
Getting actual temporary key values from the entity's tracking information failed for non-foreign key column

### ISSUE THREAD

**Issue by Nivitha-Gopalakrishnan** [dotnet/efcore#19690]
I have already reported the issue in the entity framework that we get the actual temporary key value (i.e. long negative values) stored in the table where tables have no foreign key references.

Issue Link: https://github.com/dotnet/efcore/issues/19326

In this link, they asked the sample for reproducing the issue. I have created the sample for this and kindly check it and let me know.

Source Repository Link: https://github.com/Nivitha-Gopalakrishnan/SampleProject

In this sample, I used the local database and kept it in the App_Data folder.

Using 4 tables:

1. Books
2. StudentBookMapper
3. Students
4. StudentsPersonalDetails

**Students**

Id
Name
Place
Country

Id is used as the primary key.

**Books**

Id
Name
Category
Author

Id is used as the primary key.

**StudentPersonalDetails**

Id
UserId
Department
Year

Id is used as the Primary key.  And UserId is used as a foreign key mapped to the Id column in the Students table.

**StudentBookMapper**

Id
UserId
BookId
IsPGStudent
IsActive

Id is used as the Primary key. Here, UserId is used as multiple purposes. So we didn't use the UserId as a foreign key.

**Problem:**

Here, we faced problems due to this breaking changes in EF Core 3.0 as mention here (https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-3.0/breaking-changes#temporary-key-values-are-no-longer-set-onto-entity-instances). 
However, we resolve this problem by applying for the suggested workaround solution to get the temporary key values from the below query, which returns the "Id", primary column temporary value for "Students" table

var tempIdValue = context.Entry(student).Property(e => e.Id).CurrentValue.

**Case 1:**

Students and StudentPersonalDetails tales have foreign key references relation. So the above workaround solution works fine.

_Students_

![image](https://user-images.githubusercontent.com/35961094/73050262-3b155080-3ea5-11ea-82e6-0a25cd767cb8.png)

_StudentsPersonalDetails_

![image](https://user-images.githubusercontent.com/35961094/73050312-6a2bc200-3ea5-11ea-89ee-ee91b8cead36.png)

**Case 2:**

The above workaround solution gets failed and inserted the actual temporary value (i.e. long negative numbers) in the column if they don’t have foreign key references.
Here, Students and StudentBookMapper tables have no foreign key references. So when inserting StudentId in the UserId column in the StudentBookMapper, the negative values are inserted.

_Students_

![image](https://user-images.githubusercontent.com/35961094/73050347-921b2580-3ea5-11ea-8b2f-66e0c73e1a03.png)

_StudentBookMapper_

![image](https://user-images.githubusercontent.com/35961094/73050372-b4ad3e80-3ea5-11ea-8756-01307c0f9026.png)

**Note:** We do save changes of the context once the data added to all the tables not once added to.

Kindly let me know is this an issue or anything wrong in our approach.

### Further technical details

EF Core version: 3.1
Database provider:  Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 3.1
IDE: Visual Studio 2019 16.4.1

**Comment by ajcvickers** [comment]
@Nivitha-Gopalakrishnan This is not going to work since without an FK defining the relationship EF has no way of propagating the real key value once it has been generated by the database. However, I also don't understand how this could have worked before 3.0. Can you show working code for EF Core 2.x? Then I can understand what was happening before and figure how to get that behavior in 3.1.

**Comment by Nivitha-Gopalakrishnan** [comment]
@ajcvickers  I have created a sample in EF core 2.x and find that the issue is reproduced. It seems in our previous case we have inserted a record in a table and save the changes and then inserted the value in other tables where no foreign key relationship created. Unknowing in EF core 3.x, tried the same scenario by do save changes at the end of the transaction and thought it worked in 2.x & created this issue.
